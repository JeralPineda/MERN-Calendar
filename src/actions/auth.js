import { fetchSinToken, fetchConToken } from '../helpers/fetch';
import { types } from '../types/types';

export const startLogin = (email, password) => {
   return async (dispatch) => {
      //    Petición login
      const resp = await fetchSinToken('auth', { email, password }, 'POST');

      const body = await resp.json();

      //   Si el body es ok: true guardamos el token en localstorage
      if (body.ok) {
         localStorage.setItem('x-token', body.token);
         localStorage.setItem('token-init-date', new Date().getTime()); //fecha en que se creo el token

         dispatch(
            login({
               uid: body.uid,
               name: body.name,
            })
         );
      }
   };
};

export const startChecking = () => {
   return async (dispatch) => {
      //    Petición login
      const resp = await fetchConToken('renew');

      const body = await resp.json();

      //   Si el body es ok: true guardamos el token en localstorage
      if (body.ok) {
         localStorage.setItem('x-token', body.token);
         localStorage.setItem('token-init-date', new Date().getTime()); //fecha en que se creo el token

         dispatch(
            login({
               uid: body.uid,
               name: body.name,
            })
         );
      }
   };
};

// acción para guardar el usuario autenticado en el state
const login = (user) => ({
   type: types.authLogin,
   payload: user,
});
