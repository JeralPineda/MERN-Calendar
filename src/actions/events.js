import Swal from 'sweetalert2';

import { fetchConToken } from '../helpers/fetch';
import { types } from '../types/types';

export const eventStartAddNew = (event) => {
   return async (dispatch, getState) => {
      //    Obtener info del user desde el state
      const { uid, name } = getState().auth;

      try {
         const resp = await fetchConToken('events', event, 'POST');
         const body = await resp.json();

         if (body.ok) {
            event.id = body.evento.id;
            event.user = {
               _id: uid,
               name: name,
            };

            dispatch(eventAddNew(event));
         } else {
            Swal.fire('Error', body.msg, 'error');
         }
      } catch (error) {
         console.log(error);
         Swal.fire('Error', 'No se pudo crear la nota', 'error');
      }
   };
};

export const eventStartLoading = () => {
   return async (dispatch) => {
      try {
         const resp = await fetchConToken('events');
         const body = await resp.json();

         const events = body.eventos;

         dispatch(eventLoaded(events));
      } catch (error) {
         console.log(error);
         Swal.fire('Error', 'No se encontraron notas', 'error');
      }
   };
};

const eventAddNew = (event) => ({
   type: types.eventAddNew,
   payload: event,
});

export const eventSetActive = (event) => ({
   type: types.eventSetActive,
   payload: event,
});

export const eventClearActiveEvent = () => ({
   type: types.eventClearActiveEvent,
});

export const eventUpdated = (event) => ({
   type: types.eventUpdated,
   payload: event,
});

export const eventDeleted = (event) => ({
   type: types.eventDeleted,
   payload: event,
});

const eventLoaded = (events) => ({
   type: types.eventLoaded,
   payload: events,
});
