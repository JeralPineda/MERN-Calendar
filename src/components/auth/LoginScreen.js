import React from 'react';
import { useDispatch } from 'react-redux';
import Swal from 'sweetalert2';

import { useForm } from '../../hooks/useForm';
import { startLogin, startRegister } from '../../actions/auth';

import './login.css';

export const LoginScreen = () => {
   const dispatch = useDispatch();

   const [formLoginValues, handleLoginInputChange] = useForm({
      lEmail: 'jeral@gmail.com',
      lPassword: '123456',
   });
   const { lEmail, lPassword } = formLoginValues;

   const [formRegisterValues, handleRegisterInputChange] = useForm({
      rName: 'Cristopher',
      rEmail: 'Cris@gmail.com',
      rPassword1: '123456',
      rPassword2: '123456',
   });
   const { rName, rEmail, rPassword1, rPassword2 } = formRegisterValues;

   //  Effecto en los formularios de login y registro
   const toggleForm = (e) => {
      e.preventDefault();
      const container = document.querySelector('.container');
      container.classList.toggle('active');
   };

   const handleLogin = (e) => {
      e.preventDefault();

      dispatch(startLogin(lEmail, lPassword));
   };

   const handleRegister = (e) => {
      e.preventDefault();

      //   Validación de las contraseñas
      if (rPassword1 !== rPassword2) {
         Swal.fire('Error', 'Ambas contraseñas deben de ser iguales', 'error');
      }

      dispatch(startRegister(rEmail, rPassword1, rName));
   };

   return (
      <section>
         <div className='container'>
            <div className='user signinBx'>
               <div className='imgBx'>
                  <img src='./images/5.jpg' alt='' />
               </div>

               <div className='formBx'>
                  <form onSubmit={handleLogin}>
                     <h2>Ingreso</h2>

                     <input
                        type='text' //
                        placeholder='Correo'
                        name='lEmail'
                        value={lEmail}
                        onChange={handleLoginInputChange}
                     />

                     <input
                        type='password' //
                        placeholder='Contraseña'
                        name='lPassword'
                        value={lPassword}
                        onChange={handleLoginInputChange}
                     />

                     <input type='submit' name='' value='Ingresar' />

                     <p className='signup'>
                        No tienes una cuenta ?
                        <a href='/login' onClick={toggleForm}>
                           Crear cuenta
                        </a>
                     </p>
                  </form>
               </div>
            </div>

            <div className='user signupBx'>
               <div className='formBx'>
                  <form onSubmit={handleRegister}>
                     <h2>Registro</h2>

                     <input
                        type='text' //
                        placeholder='Nombre'
                        name='rName'
                        value={rName}
                        onChange={handleRegisterInputChange}
                     />

                     <input
                        type='email' //
                        placeholder='Correo'
                        name='rEmail'
                        value={rEmail}
                        onChange={handleRegisterInputChange}
                     />

                     <input
                        type='password' //
                        placeholder='Contraseña'
                        name='rPassword1'
                        value={rPassword1}
                        onChange={handleRegisterInputChange}
                     />

                     <input
                        type='password' //
                        placeholder='Confirmar Contraseña'
                        name='rPassword2'
                        value={rPassword2}
                        onChange={handleRegisterInputChange}
                     />

                     <input
                        type='submit' //
                        name=''
                        value='Crear cuenta'
                     />

                     <p className='signup'>
                        Ya tienes una cuenta ?
                        <a href='/login' onClick={toggleForm}>
                           Ingresar
                        </a>
                     </p>
                  </form>
               </div>

               <div className='imgBx'>
                  <img src='./images/1.jpg' alt='' />
               </div>
            </div>
         </div>
      </section>
   );
};
