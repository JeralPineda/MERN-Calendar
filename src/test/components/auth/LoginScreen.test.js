import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { LoginScreen } from '../../../components/auth/LoginScreen';
import { startLogin } from '../../../actions/auth';

// mock para simular el evento startLogin
jest.mock('../../../actions/auth', () => ({
   startLogin: jest.fn(),
}));

const middleware = [thunk];
const mockStore = configureStore(middleware);

const initState = {};
let store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
   <Provider store={store}>
      <LoginScreen />
   </Provider>
);

describe('Pruebas en <LoginScreen />', () => {
   test('debe de mostrarse correctamente', () => {
      expect(wrapper).toMatchSnapshot();
   });

   test('debe de llamar el dispatch del login', () => {
      wrapper.find('input[name="lEmail"]').simulate('change', {
         target: {
            name: 'lEmail',
            value: 'jeral@gmail.com',
         },
      });

      wrapper.find('input[name="lPassword"]').simulate('change', {
         target: {
            name: 'lPassword',
            value: '123456',
         },
      });

      wrapper.find('form').at(0).prop('onSubmit')({
         preventDefault() {},
      });

      expect(startLogin).toHaveBeenCalledWith('jeral@gmail.com', '123456');
   });
});
