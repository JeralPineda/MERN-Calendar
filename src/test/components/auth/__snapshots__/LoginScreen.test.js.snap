// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <LoginScreen /> debe de mostrarse correctamente 1`] = `
<section>
  <div
    className="container"
  >
    <div
      className="user signinBx"
    >
      <div
        className="imgBx"
      >
        <img
          alt=""
          src="./images/5.jpg"
        />
      </div>
      <div
        className="formBx"
      >
        <form
          onSubmit={[Function]}
        >
          <h2>
            Ingreso
          </h2>
          <input
            name="lEmail"
            onChange={[Function]}
            placeholder="Correo"
            type="text"
            value=""
          />
          <input
            name="lPassword"
            onChange={[Function]}
            placeholder="Contraseña"
            type="password"
            value=""
          />
          <input
            name=""
            type="submit"
            value="Ingresar"
          />
          <p
            className="signup"
          >
            No tienes una cuenta ?
            <a
              href="/login"
              onClick={[Function]}
            >
              Crear cuenta
            </a>
          </p>
        </form>
      </div>
    </div>
    <div
      className="user signupBx"
    >
      <div
        className="formBx"
      >
        <form
          onSubmit={[Function]}
        >
          <h2>
            Registro
          </h2>
          <input
            name="rName"
            onChange={[Function]}
            placeholder="Nombre"
            type="text"
            value=""
          />
          <input
            name="rEmail"
            onChange={[Function]}
            placeholder="Correo"
            type="email"
            value=""
          />
          <input
            name="rPassword1"
            onChange={[Function]}
            placeholder="Contraseña"
            type="password"
            value=""
          />
          <input
            name="rPassword2"
            onChange={[Function]}
            placeholder="Confirmar Contraseña"
            type="password"
            value=""
          />
          <input
            name=""
            type="submit"
            value="Crear cuenta"
          />
          <p
            className="signup"
          >
            Ya tienes una cuenta ?
            <a
              href="/login"
              onClick={[Function]}
            >
              Ingresar
            </a>
          </p>
        </form>
      </div>
      <div
        className="imgBx"
      >
        <img
          alt=""
          src="./images/1.jpg"
        />
      </div>
    </div>
  </div>
</section>
`;
